### PR TITLE
RFC for witness multiplicities handling

### PR DESCRIPTION
**Pull request by LindaGuiga**
## Summary
This is an RFC for issue #234. It proposes a new design to handle witness multiplicities, including some changes to the common data.

Proposition for issue #234 

## Why
This is a proposition to improve the design and handling of witness multiplicities, which is currently error-prone and not very user-friendly.

## Changes
- [ ] First, I suggest having `PreprocessedColumns` handle preprocessed updates itself.
- [ ] Second, I suggest adding the traces to a `ProverData` structure, which would also contain the common data.

## Acceptance criteria
- [ ] The suggested changes are satisfactory to address the issue of handling the witness multiplicities.

## Tests
- [ ] Unit tests added/updated
- [x] Existing tests pass
- [ ] Edge cases considered (briefly note any)

## Benchmarks / Performance (if relevant)
- [ ] Before/after numbers included
- [ ] Benchmark command or link to CI perf output
- [x] No unexpected regressions in CPU / memory / proof size / compile time

## Docs (if relevant)
- [ ] Rustdoc updated
- [ ] Book/README updated
- [ ] Examples updated

## Risk
Nothing, as this PR only adds an RFC document

### CODE REVIEW

File: docs/rfcs/0001-rfc-handle-multiplicities.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+- On top of `CommonData`, we can have a `ProverData` structure which contains the common data, as well as `PreprocessedColumns` and any additional data that the prover might require. Currently, the prover regenerates the preprocessed values based on the traces. It could do the same for multiplicities, but this leads to extra overhead when the values should already have been computed beforehand. So this approach should simplify the API and very slightly improve prover performance at the expense of storing more preprocessed data before proving.
+
+### 4.2 APIs / traits / types
+
+First, we would need to add methods to `PreprocessedColumns` so it has more control over how preprocessed data is generated. I propose the following changes:
+
+```rust
+impl PreprocessedColumns {
```
Comment (Nashtare) [inline docs/rfcs/0001-rfc-handle-multiplicities.md]: nit: 
```suggestion
impl<F: Field> PreprocessedColumns<F> {
```

(note that the definition of `PreprocessedColumns<F>` does not constrain the `F` associated type but I think adding `Field` would not hurt?)

**Review by Nashtare — COMMENTED** [review]
It's a nice proposal! Alternatively we could possibly "embed" the preprocessed info into the AIRs / structs themselves, to make the logic more visible on each instance and have `generate_preprocessed_columns` fully agnostic but this may be overkill so I'd be happy with the current approach too.

File: docs/rfcs/0001-rfc-handle-multiplicities.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+            ConstAir::<Val<SC>, D>::trace_to_matrix(&traces.const_trace);
+
+        // Apply similar changes to the rest of the primitive and non-primitive operations.
+        ..
+    }
+```
+
+With this approach, we can also remove `trace_to_preprocessed` in the various airs where it is implemented. Currently, the method is a bit redundant with `generate_preprocessed` since it also generates preprocessed data based on the traces. This redundancy makes it error-prone, so being able to get rid of it is, in my opinion, another benefit of this approach.
+
+Note that ideally, we should also change `CommonData` in Plonky3, as it contains some prover data in `GlobalPreprocessed`. And so, on the Plonky3 side, we could remove `prover_data` from `GlobalPreprocessed` and instead store it in a `ProverOnlyData` which also contains `CommonData`.
```
[the remark this answers is not in the document]
Comment (LindaGuiga) [reply docs/rfcs/0001-rfc-handle-multiplicities.md]: Indeed, I didn't mean it as a blocker: I just wanted to mention it. I will add an issue on P3.

File: docs/rfcs/0001-rfc-handle-multiplicities.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    /// with `wid`'s witness index, and updates the witness multiplicity.
+    fn register_non_primitive_witness_read(&mut self, op_type: NonPrimitiveOpType, wid: WitnessId) {
+        self.register_non_primitive_witness_reads(op_type, &[wid]);
+    }
+
+    /// Extends the preprocessed data of the `table_idx`-th primitive operation 
+    /// with `values`.
+    fn register_primitive_preprocessed_no_read(&mut self, table_idx: usize, values: &[F]) {
+        self.primitive[table_idx].extend(values);
+    }
```
[the remark this answers is not in the document]
Comment (LindaGuiga) [reply docs/rfcs/0001-rfc-handle-multiplicities.md]: I made all the methods return a `CircuitError`, and added a safeguard for this and for the case where `primitive` does not have the right length.
